tracing/optimisation/utils.py

```diff
@@ -0,0 +1,66 @@
+from dataclasses import dataclass
+from traceback import linecache # type: ignore
+import functools
+import io
+import typing
+import tokenize
+
+
+# Start with characters or underscore, rest is the same plus numbers
+@dataclass
+class FrameWithMetadata:
```

Comment:
Isn't it more clean to define a function get_frame_metadata_from(frame) -> FrameMetaData, which contains the properties?

Reply:
The optimisations need both due to hidden mutable references. This cost me three days of debugging, which is why I put everything in this class